src/lib/check-in.ts

```diff
@@ -1,14 +1,85 @@
 import console from 'console';
-import HttpStatus from 'http-status';
 import * as Luxon from 'luxon';
 import { Reservation } from '../lib/reservation';
 import * as SwClient from '../lib/sw-client';
 
-export function makeFetchCheckinDataAttempts(
-  reservation: Reservation,
-  headers: Record<string, string>,
-  logger?: typeof console
-) {
+export interface MakeFetchCheckinDataAttemptsInput {
+  reservation: Reservation;
+  headers: Record<string, string>;
+  attemptLimit: number;
+  millisPerRequest: number;
+  logger?: typeof console;
+}
+
+/**
+ * Make many requests in attempt to fetch checkin data as quickly as possible.
+ *
+ * Does not wait for each request to complete. Instead simply makes requests on an interval until
+ * one is successful. Thus, we will likely have multiple requests in flight at once.
+ */
+export async function makeFetchCheckinDataAttempts(input: MakeFetchCheckinDataAttemptsInput) {
+  // Create a cancel function that can be used to cancel the attempts. We are essentially
+  // implementing a cancelable promise. We use this to cancel the remaining attempts once we get one
+  // successful response.
+
+  const signal = createCancellableSignal();
+
+  // schedule requests such that they are made on each `input.millisPerRequest` interval
+
+  const promises = [];
+  for (let attempt = 0; attempt < input.attemptLimit; attempt++) {
```

Comment:
This is pretty clever! I'm still getting familiar with node/javascript concepts, but is this what you're essentially doing?

Creating 80 requests with a delay on when to actually execute (so the first one waits 0 seconds, next 0.25, next 0.5), once you get a successful response from any of them, cancel all request that are still waiting for their time.

What I'm a little confused on, how does this piece work? `const response = await Promise.any(promises);`

More specifically, even in the case of a "failed" request (in `makeDelayedRequest`), we'll catch and log the error, but then we'll still return `response` which would be empty since the request failed. This would trigger our `const response = await Promise.any(promises);` to trigger and we'll cancel all the rest of the request we make.

Or am I misunderstanding something?

Reply:
You are exactly correct